### ISSUE TITLE
How to read metrics stored in mlflow?

### ISSUE THREAD

**Issue by nsidn98** [mlflow/mlflow#2907]
Is it possible to read the metrics stored in the mlflow board as a numpy array or something else? I want to perform pythonic operations on them.

**Comment by nsidn98** [comment]
Hi, Thanks for the link to the blog. The code in the blog seems to give an array of the final values of the metrics. Is it possible to get an array of metrics stored in each of the experiment?

For example, I want to plot the training_loss of each experiment separately with respect to the epochs.

**Comment by mberr** [comment]
Hi @dmatrix , @nsidn98 . 

Is there any progress towards this issue? I am also interested in this use case. To be a bit more specific, I am looking for a method with an interface similar to

```python
get_metrics_for_run(run_uuid: str, metrics: Optional[List[str]] = None) -> pandas.DataFrame
```

returning a dataframe (in long-format) with columns `timestamp, step, metric, value` comprising all metric values logged throughout a run.


Moreover, it may be interesting to retrieve these dataframes for all runs associated with an experiment, returning a dataframe with an additional `run_uuid` column, which can be joined with the dataframe containing the runs' parameters, e.g. obtained from `mlflow.search_runs(experiment_ids=experiment_id)`.

```python
get_metrics_for_experiment(experiment_uuid, metrics: Optional[List[str]] = None) -> pandas.DataFrame
```

Providing the metrics of interest beforehand allows more efficient backend operations than first retrieving all metrics and then filtering on the resulting dataframe.

Having access to this data allows for instance analysing training progress similar to what is shown here ( https://github.com/mlflow/mlflow/issues/2768#issuecomment-621568259 ) outside of the MLFlow web interface.

**Comment by AveshCSingh** [comment]
@nsidn98, I believe you can use the get_metric_history function mentioned by @mberr to get an array of metrics for a run, and cal this repeatedly for every run in your experiment.

I've attached the label "priority/awaiting-more-evidence" to this issue, because I don't currently see a lot of value in adding a batch API here.

**Comment by Demetrio92** [comment]
Following the above discussion I implemented a simple metrics to pandas routine: 

```python
def get_metric_names(mlflow_client, run_id):
    run_data = mlflow_client.get_run(run_id).data.to_dictionary()
    metric_names = list(run_data['metrics'].keys())
    return metric_names


def pull_run_metrics_as_df(mlflow_client, run_id, metric_names=None):
    if metric_names is None:
        metric_names = get_metric_names(mlflow_client, run_id)

    for_pd_collect = []
    for metric in metric_names:
        metric_history = mlflow_client.get_metric_history(run_id=run_id, key=metric)
        pd_convertible_metric_history = [
            {
                'metric_name': mm.key,
                'step': mm.step,
                'timestamp': mm.timestamp,
                'value': mm.value,
            }
            for mm in metric_history
        ]
        for_pd_collect += pd_convertible_metric_history

    metrics_df = pd.DataFrame.from_records(for_pd_collect)
    return metrics_df
```

beware, there is a constant
`GET_METRIC_HISTORY_MAX_RESULTS = 25000`, so once you're over the limit you won't get full results.

**Comment by Demetrio92** [comment]
@AveshCSingh @dmatrix 

For a moderately-sized RL project we decided to go with MLFlow. Over a training run with 240k steps we track some 200 metrics. For 10 runs we get approximately 1_000_000 data points. 

If using the code I provided above to pull this data into a pandas dataframe for further analysis, it takes 6 seconds per run, or a minute for 10 runs. I am working with a local 1Gbit network. But, in fact, it takes the same amount of time even when this code is run directly on the same machine that hosts mlflow. This does not scale. 

The biggest bottleneck is `mlflow_client.get_metric_history`, although iterating through every value just to be able to put it into a dataframe isn't pretty either. 

If I stage this data locally, I get **10mb of parquet files**. So: 1 minute of delay for 10 mb of tracking data. 

-------

My possible next steps are 
1) implement a data puller on the server side that will directly connect to the storage without any `PagedHistory`
2) implement a local caching logic
3) track metrics in pandas and send them to the mlflow server as an artifcat

I am convinced that neither should be necessary and it would make sense to look into ways of more efficiently providing metrics history to the users, where a pandas dataframe as an output would also be a user-friendly output instead of a `PagedList[Metric]`.
